pkg/capi/components/CCVariables/index.vue

```diff
@@ -0,0 +1,251 @@
+      this.$nextTick(() => {
+        this.rerenderKey = randomStr();
+      });
+    },
+  },
+
+  created() {
+    this.updateVariableDefaults(this.variableDefinitions, []);
+  },
+
+  computed: {
+    variableDefinitions() {
+      const allVariableDefinitions = this.clusterClass?.spec?.variables || [];
+
+      if (!this.machineDeploymentClass && !this.machinePoolClass) {
+        return allVariableDefinitions;
+      }
+      const variableNames = this.machineScopedJsonPatches.reduce((names, patch) => {
+        const valueFromVariable = patch?.valueFrom?.variable;
+
+        if (!valueFromVariable) {
+          return names;
+        }
+
+        // the value here could be a field or index of the variable, defined <variable definition.name>.<some field> or <variable definition name>[i]
+        const parsedName = valueFromVariable.split(/\.|\[/)[0];
+
+        if (parsedName !== 'builtin') {
+          names.push(parsedName);
+        }
+
+        return names;
+      }, []);
+
+      return allVariableDefinitions.filter((v: ClusterClassVariable) => variableNames.includes(v.name));
+    },
+
+    machineScopedJsonPatches() {
+      if (!this.machineDeploymentClass && !this.machinePoolClass) {
+        return [];
+      }
+      const out = [] as any[];
+      const matchName = this.machineDeploymentClass || this.machinePoolClass;
+      const matchKey = this.machineDeploymentClass ? 'machineDeploymentClass' : 'machinePoolClass';
+
+      const patches = this.clusterClass?.spec?.patches || [];
+
+      patches.forEach((p) => {
+        const definitions = p?.definitions || [];
+
+        definitions.forEach((definition: any) => {
+          const matchMachines = definition?.selector?.matchResources?.[matchKey]?.names || [];
+
+          if (matchMachines.includes(matchName)) {
+            out.push(...definition.jsonPatches);
+          }
+        });
+      });
+
+      return out;
+    },
+  },
+
+  methods: {
+    valueFor(variableDef: ClusterClassVariable) {
+      return (this.value.find((variable: any) => variable.name === variableDef.name) || {}).value;
+    },
+
+    updateVariables(val: any, variableDef: ClusterClassVariable) {
+      const out = [...this.value] as Array<any>;
+      const existingIdx = this.value.findIndex((variable: any) => variable.name === variableDef.name);
+
+      if (existingIdx >= 0) {
+        out[existingIdx].value = val;
+      } else {
+        out.push({ value: val, name: variableDef.name });
+      }
+      this.$emit('input', out);
+    },
+
+    /**
+   * When the cluster class changes, update the variables array:
+   *  remove any variables not defined in the new cluster class
+   *  if a variable is defined in both cluster classes, clear out the old default
+   *  set default values from the new cluster class definitions
+   *  if a variable is defined in both old and new cluster classes, and the user has set its value to something other than the old default, preserve that
```

Comment:
Are you sure we want to preserve it? I'd reset it to the default

Reply:
I mean I guess it depends on why a user would be changing the cluster class after entering in variables. I would definitely find it tiresome to re-enter data if I was swapping between cluster classes to verify that I had the one I wanted.

Reply:
Fair.